docs/aot.md

```diff
@@ -0,0 +1,232 @@
+# Ahead-of-time lowering and compilation
+
+JAX offers several transformations, such as `jax.jit` and `jax.pmap`, returning
+a function that is compiled and runs on accelerators or the CPU. As the JIT
+acronym indicates, all compilation happens _just-in-time_ for execution.
+
+Some situations call for _ahead-of-time_ (AOT) compilation instead. When you
+want to fully compile prior to execution time, or you want control over when
+different parts of the compilation process take place, JAX has some options for
+you.
+
+First, let's review the stages of compilation. When a function `f` output by
+{func}`jax.jit` is invoked with arguments, say `f(x, y)` where `x` and `y` are
+arrays, JAX does the following in order:
+
+1. **Specialize** `f` to properties of the arguments `x` and `y` (usually their
+   shape and element type).
+
+2. **Lower** this specialization to the XLA compiler's input language, HLO.
+
+3. **Compile** the lowered HLO program to produce an optimized executable for
+   the target device (CPU, GPU, or TPU).
+
+4. **Execute** the compiled executable with the arrays `x` and `y` as arguments.
+
+JAX's AOT API gives you direct control over steps #2, #3, and #4 (but [not
+#1](#inspecting-specializations)), plus some other features along the way. An
+example:
+
+```python
+>>> import jax
+>>> import jax.numpy as jnp
+>>> import numpy as np
+
+>>> def f(x, y): return 2 * x + y
+>>> x, y = 3, 4
+
+>>> lowered = jax.jit(f).lower(x, y)
+
+>>> # Print lowered HLO
+>>> print(lowered.as_text())
+module @jit_f.0 {
+  func.func public @main(%arg0: tensor<i32>, %arg1: tensor<i32>) -> tensor<i32> {
+    %0 = mhlo.constant dense<2> : tensor<i32>
+    %1 = mhlo.multiply %0, %arg0 : tensor<i32>
+    %2 = mhlo.add %1, %arg1 : tensor<i32>
+    return %2 : tensor<i32>
+  }
+}
+
+>>> compiled = lowered.compile()
+
+>>> # Query for cost analysis, print FLOP estimate
+>>> compiled.cost_analysis()[0]['flops']
+2.0
+
+>>> # Execute the compiled function!
+>>> compiled(x, y)
+DeviceArray(10, dtype=int32)
+```
+
+See the {mod}`jax.stages` documentation for more details on what functionality
+the lowering and compiled functions provide.
+
+In place of `jax.jit` above, you can also `lower(...)` the result of
+{func}`jax.pmap`, as well as `pjit` and `xmap` (from
+{mod}`jax.experimental.pjit` and {mod}`jax.experimental.maps` respectively). In
+each case, you can `compile()` the result similarly.
+
+All optional arguments to `jit`---such as `static_argnums` and `device`---are
```

Comment:
FYI I think we want to deprecate and remove `device` soon, so we might want to avoid referring to it (so as to have less to update later).